Add fio2_percent and oxygen_flow_rate

pipliggins opened the pull request.

ekamau reviewed (changes requested):
thanks @pipliggins - lots of work!
all look very good; and just a few minor changes to look at ..

pipliggins commented on isaric/parsers/idor-brazil.toml:
`works_lab` is a super type, which includes people who work in an (unspecified) laboratory as well as a microbiology lab. There are therefore scenarios like in idol-brazil where both `works_lab` and `works_microbiology_lab` are present, and others where only `works_lab` is present as a _microbiology_ lab has not been specified in the data dictionary.

pipliggins commented on isaric/parsers/sprintsari-australia.toml:
```diff
@@ -109,6 +109,11 @@
+  [subject.works_lab]
+    field = "healthwork_erterm"
+    description = "Employed a a healthcare worker or any of following professions?"
+    values = { 1 = false, 2 = false, 3 = false, 4 = true, 5 = false, 7 = false, 8 = false, 9 = false, 10 = false, 11 = false }
```
healthwork_eterm = 4 relates to a microbiology lab worker, and therefore should flag both `works_lab` and `works_microbiology_lab` as True (see below)

pipliggins commented on isaric/parsers/ccp-guinea.toml:
```diff
@@ -578,30 +573,29 @@
 [visit.treatment_high_flow_nasal_cannula]
   field = "overall_oxygen_cmtype___2"
   description = "Oxygen interface type - HF nasal cannula"
   values = { 2 = true }
```
thanks for catching that, the overall_oxygen stuff should have been removed.

pipliggins commented:
@sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.

abhidg commented on isaric/parsers/isaric-ecmo.toml:
```diff
@@ -1116,6 +1121,18 @@
+  name = "fio2_percent"
+  phase = "study"
+  date = { ref = "dailyDateHierarchy" }
+  value = { field = "daily_fio2b_lborres", apply = { function = "Percentage" } }
```
what is the difference between fio2 and fio2b?

pipliggins replied:
fio2b contains values as a percentage, fio2 as decimals (or possible vv.) - but both appear to contain a mix of decimals and percentage values in practice

sadiekelly replied:
can we create a function to convert to percentage only if the value is between 0.21 and 1.0? and leave as is if between 21 and 100 (%). There are probably values in between 1 and 21 that will have to remain as reported. Would that be possible, as I don't think there's ambiguity in the units if the values are within those clearly defined ranges

pipliggins commented on isaric/parsers/ccp-cameroon.toml:
```diff
@@ -879,7 +1058,10 @@
   date = { field = "q0_signs_symptoms" }
-  if.all = [{redcap_event_name = {"=~" = 'm*\d_arm_1'}}, {q24_signs_symptoms = {"!=" = ""}}]
+  if.all = [
+    { redcap_event_name = { "=~" = 'm*\d_arm_1' } },
```
I think we want it to match m3/m6/m9, looking at the data

abhidg replied:
^ then the regex can be 'm\d_arm_1'

pipliggins replied:
Thanks @abhidg, I've updated this

sadiekelly commented on isaric/parsers/ccp-drc.toml:
```diff
@@ -298,6 +298,11 @@
 
+[subject.works_lab]
+  field = "microb"
+  description = "Employee of a microbiology laboratory"
```
@pipliggins I think for creating the works_lab supertype we'd only map the Y responses if the question specifically asks about working in a microbiology lab, as someone could respond 'N' to not working in a microbiology lab but they are still a lab worker, what do you think?

pipliggins replied:
@sadiekelly so you mean if a question asks 'are you a microbiology worker?' we'd only map the yes responses to works_lab, but both Y and N for works_microbiology_lab? Yes, I think that makes sense. We might potentially need to consider this logic for other super types as well...

sadiekelly replied:
yes, that's right, I think that's the best way. Which of the other super types are you thinking of?

pipliggins replied:
I haven't thought it through... but possibly for things like oxygen therapy - @abhidg remind me, if a 'false' response is given within combinedType = any, will that override a 'True' response from earlier?

sadiekelly commented on isaric/parsers/ccp-ghana.toml:
```diff
@@ -1332,6 +1336,12 @@
+  name = "fio2_percent"
+  phase = "study"
+  date = { field = "vitals_date" }
+  value = { field = "fi02", apply = { function = "Percentage" } }
```
@pipliggins just to check, the function "Percentage" converts the fractional values (0.21-1.0) to percentage (21-100%)?

pipliggins replied:
Yes! It will also leave any existing percentages alone, as many datasets seem to have a mix of both regardless of the supposed unit.

sadiekelly commented on isaric/parsers/ccp-senegal.toml:
```diff
@@ -244,8 +244,14 @@
+[subject.works_lab]
+  field = "labwork_erterm"
+  description = "employed in a biological analysis laboratory?"
+  ref = "Y/N/NK"
```
@pipliggins I'm thinking perhaps this variable should only map to works_lab, as it is not specifically asking about micro lab work

sadiekelly commented on isaric/parsers/idor-brazil.toml:
```diff
@@ -267,6 +267,10 @@
+[subject.works_lab]
+  field = "labwork_erterm"
+  ref = "Y/N/NK"
+
```
same comment for mapping only Y to works_lab where the question asks specifically about microbiology lab work

sadiekelly commented on isaric/parsers/isaric-ccpuk.toml:
```diff
@@ -80,6 +80,11 @@
+  [subject.works_lab]
+    field = "labwork_erterm"
+    ref = "Y/N/NK"
+    description = "Employed in a microbiology laboratory?"
```
same comment mapping Y responses only to works_lab from the microbiology question

sadiekelly commented on isaric/parsers/isaric-ccpuk.toml:
```diff
@@ -1020,6 +1025,18 @@
+  name = "fio2_percent"
+  phase = "study"
+  date = { field = "daily_dsstdat", description = "Date of assessment" }
+  value = { field = "daily_fio2b_lborres", apply = { function = "Percentage" } }
```
daily_fio2b_lborres is already provided as a percentage, so is the "Percentage" function not needed here?
daily_fio2_lborres is provided as a fraction of 1 so the percentage function looks ok for this one

sadiekelly commented on isaric/parsers/isaric-ecmo.toml:
```diff
@@ -274,6 +274,11 @@
+    field = "labwork_erterm"
+    ref = "Y/N/NK"
+    description = "Employed in a microbiology laboratory?"
+
```
same comment for mapping Y responses from microbiology question

sadiekelly commented on isaric/parsers/isaric-rapid.toml:
```diff
@@ -1538,6 +1538,18 @@
+  phase = "study"
+  date = { field = "ccm_a_dat" }
+  value = { field = "ccm_a_fio2_lborres", apply = { function = "Percentage" } }
+
```
same comment for applying "Percentage" function here. 
Fi02 can also be captured in L/min, and an approximate conversion can be made from this to % Fi02, but it is dependent on the method of oxygen therapy. Should we create an fio2_Lmin attribute to capture this, then users can decide for themselves how/if they wish to convert the values?

sadiekelly commented on isaric/parsers/ncov-malaysia.toml:
```diff
@@ -126,6 +126,11 @@
+[subject.works_lab]
+  field = "lab_worker"
+  description = "Employed in a microbiology laboratory?"
+  ref = "Y/N"
```
same comment for mapping to works_lab from microbiology question

sadiekelly commented on isaric/parsers/ncov-malaysia.toml:
```diff
@@ -881,6 +886,28 @@
+  ]
+  date = { field = "daily_assess_date" }
+  value = { field = "daily_fio2_data", apply = { function = "Percentage" } }
+
```
there's also Fio2 data in the variable fio_prone, daily_assess_date can be used

sadiekelly commented on isaric/parsers/ccp-guinea.toml:
```diff
@@ -578,43 +560,38 @@
   excludeWhen = "none"
   fields = [
-    { field = "dexamethasone", values = { 1 = "Dexamthasone" } },
     { field = "hydrocortisone", values = { 1 = "Hydrocortisone" } },
```
@pipliggins I'd noted from my initial review to use daily_corticost_cmtrt___1 = 1 to map to the visit attribute visit.treatment_dexamethasone. Maybe this now comes under this corticosteroid_type attribute? If so, daily_corticost_cmtrt___1 (dexa), ___2 (hydrocortisone), ___3 (prednisolone), and ___4 (methylprednisolone) can be mapped here also

sadiekelly commented on isaric/parsers/ccp-guinea.toml:
```diff
@@ -659,128 +636,140 @@
   field = "overall_oxygen_cmoccur"
   description = "Oxygen therapy?"
-  ref = "Y/N/NK"
+  ref = "Y/N"
```
to add other oxygen therapy variables to this attribute (non-inv, invasive etc.), perhaps I'm not seeing those in this amendment

sadiekelly commented:
> @sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.

@pipliggins thank you! looks ok - I just added a comment regarding the oxygen_therapy supertype to include all variables that indicate any oxygen therapy was given. I noted from my review the addition of mapping palpitations and edema but these attributes don't seem to be present in the version of the parser I'm looking at - are these attributes still being mapped?

pipliggins commented:
> > @sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.
> 
> @pipliggins thank you! looks ok - I just added a comment regarding the oxygen_therapy supertype to include all variables that indicate any oxygen therapy was given. I noted from my review the addition of mapping palpitations and edema but these attributes don't seem to be present in the version of the parser I'm looking at - are these attributes still being mapped?

Thanks Sadie - these two attributes certainly aren't currently being mapped, I'm not actually sure they ever were? I've added those Guinea changes here, it's bit confusing with changes being made to the same parts of the same parser on two different branches!

pipliggins commented:
Leaving this open until the difference between fio2 and o2_flow is resolved (should O2_flow be mapped as fio2_Lmin?) and we've decided whether fio2 should be included in the oxygen_therapy super type.

sadiekelly commented:
> > > @sadiekelly would you mind checking the guinea parser particularly? It seems like it was merged before your review changes were applied, so it's worth going back and double checking I caught everything.
> > 
> > 
> > @pipliggins thank you! looks ok - I just added a comment regarding the oxygen_therapy supertype to include all variables that indicate any oxygen therapy was given. I noted from my review the addition of mapping palpitations and edema but these attributes don't seem to be present in the version of the parser I'm looking at - are these attributes still being mapped?
> 
> Thanks Sadie - these two attributes certainly aren't currently being mapped, I'm not actually sure they ever were? I've added those Guinea changes here, it's bit confusing with changes being made to the same parts of the same parser on two different branches!

that's ok, the oxygen_therapy additions look ok, I guess pending any FiO2. I'm not sure where those attributes came from, but will gladly ignore!

sadiekelly commented:
> Leaving this open until the difference between fio2 and o2_flow is resolved (should O2_flow be mapped as fio2_Lmin?) and we've decided whether fio2 should be included in the oxygen_therapy super type.

@pipliggins I've spoken with Laura and Felipe about these queries and have some answers, as below;

Regarding including FiO2 in the oxygen_therapy supertype, this can be done as any indication of supplemental oxygen in a boolean question can be considered as oxygen therapy. If we don't have the boolean question and just a value for FiO2 in either % or fraction of 1, then it should only be mapped to oxygen therapy where the value is greater than 0.21 or 21 % (these entries would be considered data entry errors as they indicate oxygen as at room air rather than supplemental treatment).

FiO2 captured in L/min is equivalent to other variables capturing oxygen flow volume in L/min so these can be combined into one attribute such as oxygen_flow_volume. FiO2 in L/min is not really an FiO2 measurement but can be converted, the conversion depending on the method of oxygen therapy used. So I think best to label these as oxygen flow volume rather than FiO2. We may need to consider adding a context note where we have different variables capturing oxygen flow volume relating to different oxygen therapy methods e.g. invasive, noninvasive as you picked up in the french database

pipliggins commented:
@sadiekelly with the 2 new definitions here (oxygen_flow_rate and fio2) - do you think the presence of either/both should be used as an indicator of oxygen therapy in the visit table? Or just one?

pipliggins commented:
Hi @sadiekelly, would you mind doing a final check through this? I think I've done everything we've discussed, but a second pair of eyes before I merge would be great!

sadiekelly commented on isaric/parsers/ccp-ghana.toml:
```diff
@@ -883,37 +888,34 @@
+      { field = "overall_oxygen_cmtype___3", values = { 1 = "Mask" } },
+      { field = "overall_oxygen_cmtype___2", values = { 1 = "HF nasal cannula" } },
+      { field = "overall_oxygen_cmtype___1", values = { 1 = "Nasal prongs" } },
     ]
```
the overall cmtype and cmdose variables are both checkboxes, so if there is more than one option selected for each then it is not possible to associate the type with the dose given of that type. I think it's useful to associate type and L/min as this enables calculation of FiO2, but I think we can only do this where one dose and one type is indicated

sadiekelly replied:
daily variables are single select dropdown so they can be associated

sadiekelly commented on isaric/parsers/cocoa_pediatrics.toml:
```diff
@@ -559,12 +561,12 @@
   duration_type = "event"
-  context = ["over entire visit"]
+  context = ["Maximum flow rate over entire visit"]
 
```
can we add context nasal prongs or mask from xygn_thrpintfc_trt_v3?

sadiekelly commented on isaric/parsers/ccp-netherlands.toml:
```diff
@@ -310,6 +310,7 @@
       { field = "patient_interventions#Extra_corporeal_life_support_ECLS", values = { 1 = true, 0 = false } },
       { field = "Extracorporeal_support", ref = "Y/N" },
+      { field = "fio2_1", if = { fio2_1 = { ">" = 21 } }, apply = { function = "isNotNull" } },
     ]
```
the variable signssymp_L_oxygen also gives L oxygen, I think at admission although not super clear! This can also  be mapped to the oxygen flow attribute with description 'amount of oxygen therapy (L)'

sadiekelly commented on isaric/parsers/ccp-guinea.toml:
```diff
@@ -662,6 +662,9 @@
+    { field = "fio2", if = { fio2 = {">"= 21}}, apply = {function = "isNotNull"}},
+    { field = "any_supplemental_oxygen_value", if = { fio2_0_21_1_0_unit = 2 }, apply = {function = "isNotNull"} },
+    { field = "any_supplemental_oxygen_value", if.all = [{ fio2_0_21_1_0_unit = 1 }, { any_supplemental_oxygen_value = { ">" = 21}}], apply = {function = "isNotNull"} }
   ]
```
daily_oxygen_cmdose_1 can be added to the oxygen_therapy attribute

sadiekelly commented on isaric/parsers/ccp-guinea.toml:
```diff
@@ -1391,6 +1394,36 @@
+    excludeWhen = "none"
+    fields = [
+      { field = "daily_oxygen_cmtype_1", values = { 1 = "nasal cannula", 2 = "HFNC", 3 = "mask", 4 = "mask with reservoir", 5 = "CPAP/NIV mask", 6 = "Unknown" } },
+    ]
```
add daily_oxygen_cmdose_1

sadiekelly commented on isaric/parsers/isaric-core.toml:
```diff
@@ -876,6 +880,23 @@
+  [[visit.treatment_oxygen_therapy.fields]]
+    field = "daily_fio2c_lborres"
+    apply = { function = "isNotNull" }
+
```
oxygen_cmdose can be added to the oxygen.therapy fields. Probably already covered by oxygen_cmoccur as this is the associated dose, but just in case

sadiekelly commented on isaric/parsers/isaric-core.toml:
```diff
@@ -1981,3 +2002,10 @@
+  phase = "study"
+  date = { field = "daily_dsstdat" }
+  value = { field = "daily_fio2c_lborres" }
+  context = ["Maximum value over 24hrs"]
```
oxygen_cmdose can be added to oxygen_flow_rate. This is the maximum flow volume.

sadiekelly commented on isaric/parsers/isaric-ecmo.toml:
```diff
@@ -1168,6 +1174,49 @@
+        { daily_ecmo_prtype = 4 },
+      ] } },
+      { field = "daily_ecmo_prtype", values = { 1 = "VV ECMO", 2 = "AV ECMO", 3 = "Central ECMO" } },
+    ]
```
i think it might be best to drop the context here, as this data doesn't appear to have been collected in direct relation to the fio2 values, so we don't know whether the fio2 value given relates to one or any of the oxygen therapy types

pipliggins replied:
Do you mean the whole context section, or just the daily_ecmo_prtrt section?

sadiekelly replied:
@pipliggins I think the whole context section should be dropped as no specific device was indicated alongside the FiO2 values

sadiekelly commented on isaric/parsers/isaric-ccpuk.toml:
```diff
@@ -1077,6 +1083,34 @@
+      { field = "daily_noninvasive_prtrt", values = { 1 = "non-invasive" } },
+      { field = "daily_invasive_prtrt", values = { 1 = "invasive" } },
+      { field = "daily_ecmo_prtrt", values = { 1 = "ECMO" } },
+    ]
```
i think it might be best to drop the context values here as I don't think the variables are explicitly related and we don't know whether the fio2 value is related to one of these interventions

sadiekelly commented on isaric/parsers/isaric-rapid.toml:
```diff
@@ -1815,6 +1821,40 @@
+      { field = "ccm_b_noninvasive_prtype___2", values = { 1 = "CPAP" } },
+      { field = "ccm_b_noninvasive_prtype___3", values = { 1 = "Other non-invasive" } },
+      { field = "ccm_b_invasive_prtrt", values = { 1 = "invasive mechanical" } },
+    ]
```
I think best to drop context, as above

sadiekelly commented on isaric/parsers/ncov-france.toml:
```diff
@@ -466,14 +466,35 @@
     fields = [
       { field = "a_o2_on", description = "O2 flow (L/min), nasal O2" },
       { field = "a_of_o2", description = "O2 flow (L/min), HFNC" },
+      { field = "a_vni_ti", description = "Inspiratory trigger (L/min)" },
```
I'm not sure inspiratory trigger can be used for oxygen flow, seems like a setting on the NIV equipment

sadiekelly commented:
> Hi @sadiekelly, would you mind doing a final check through this? I think I've done everything we've discussed, but a second pair of eyes before I merge would be great!

hi @pipliggins! I've added a few comments on the updated files - let me know if something is not clear - thanks!
